src/commons/crypto/signing/dispatch/signerrouter.rs

```diff
@@ -0,0 +1,676 @@
+/// standby in a "pending" set until we can verify that they are reachable and usable and can determine which
+/// [SignerMapper] [Handle] to assign to them.
+///
+/// The Krill configuration file (TODO) defines named signers with a type (openssl, kmip or pkcs#11) and type specific
+/// settings (key dir path, hostname, port number, TLS certificate paths, username, password, slot id, etc) and assigns
+/// signers a role (default signer, one-off signer, rand signer, keyroll signer, etc) either explicitly or by default.
+///
+/// Keys created using signers in a previous Krill process MUST have been registered by the signer with the
+/// [SignerMapper] to indicate that the signer owns/possesses the key, and how to map from the [KeyIdentifier] to any
+/// internal signer specific key id. When a new Krill process starts it will need to know for any given [KeyIdentifier]
+/// which signer that was created should be used to work with the key. Rather than rely on operator supplied signer
+/// names being stable or requiring operators to also maintain a stable signer id in the config, we instead "bind" the
+/// signer backend to the signer handle that owns the set of keys stored in the [SignerMapper].
+///
+/// Binding is done by asking the signer on first use to create a new key pair for which we save the public key and the
+/// signer specific internal private key identifier and combine them into a unique [Handle] for use by the signer with
+/// the [SignerMapper]. We also store some metadata about the signer backend with the [Handle] in the [SignerMapper]
+/// which allows us to see if the configuration and/or backend properties change over time.
+///
+/// On subsequent bindings we determine which signer maps to which [SignerMapper] [Handle] by extracting the private key
+/// signer specific internal id from the [Handle] and asking each signer to sign a challenge using that key. We then
+/// verify the signature using the saved public key. If the signer doesn't know the internal private key id or produces
+/// an incorrect signature we know that the signer doesn't possess the binding key and thus likely isn't the signer we
+/// should go to for the keys mapped to the [SignerMapper] [Handle] corresponding to the binding key.
+///
+/// By binding this way we both verify that the signer is usable (at least for key pair creation and signing) and that
+/// we are using a signer that should have the keys we expect it to possess.
+///
+#[cfg(feature = "hsm")]
+enum IdentifyResult {
+    Unavailable,
+    Corrupt,
+    Identified(Handle),
+    Unusable,
+    Unidentified,
+}
+
+#[cfg(feature = "hsm")]
+enum RegisterResult {
+    NotReady,
+    ReadyVerified(Handle),
+    ReadyUnusable,
+}
+
+#[cfg(not(feature = "hsm"))]
+impl SignerRouter {
+    fn bind_ready_signers(&self) {}
+}
+
+#[cfg(feature = "hsm")]
+impl SignerRouter {
+    fn bind_ready_signers(&self) {
```

Comment:
Right, so after digging around the code a bit I think that the is used this way and called at runtime to deal with HSMs that may not be available at startup - and that furthermore this reports an error, but does not return with an error, to prevent that one failing signer makes all other signers unavailable.

Perhaps add a documentation comment to this effect to this fn?

Reply:
Does this help?

Reply:
created an issue, we can do this later: https://github.com/NLnetLabs/krill/issues/729